xarray/core/indexing.py

```diff
@@ -654,6 +686,185 @@ def _outer_to_numpy_indexer(key, shape):
         return _outer_to_vectorized_indexer(key, shape)
 
 
+def decompose_indexer(indexer, shape, mode):
+    if isinstance(indexer, VectorizedIndexer):
+        return _decompose_vectorized_indexer(indexer, shape, mode)
+    if isinstance(indexer, OuterIndexer):
+        return _decompose_outer_indexer(indexer, shape, mode)
+    if isinstance(indexer, BasicIndexer):
+        return indexer, ()
+    raise TypeError('unexpected key type: {}'.format(indexer))
+
+
+def _decompose_vectorized_indexer(indexer, shape, mode):
+    """
+    Decompose vectorized indexer to the successive two indexers, where the
+    first indexer will be used to index backend arrays, while the second one
+    is used to index loaded on-memory np.ndarray.
+
+    Parameters
+    ----------
+    indexer: VectorizedIndexer
+    mode: str, one of ['basic' | 'outer' | 'outer_1vector' | 'vectorized']
```

Comment:
This is a minor point, but instead of using strings for modes (which can be error prone), it would be better to use objects.

This is a perfect use-case for [enums](https://docs.python.org/3/library/enum.html), but it's probably not worth adding a new dependency just for this. Still, we can use objects here, e.g.,
```python
# like enums, the values for these class variables are arbitrary
class IndexingSupport(object):  # could inherit from enum.Enum on Python 3
    BASIC = 'BASIC'
    OUTER = 'OUTER'
    OUTER_1VECTOR = 'OUTER_1VECTOR'
    VECTORIZED = 'VECTORIZED'
```